bento-platform/bento_lib
# Features/search in

This PR adds new operators (`#in`, `#list`) to the query language, to allow searching for field values matching a list of values (an equivalent to the `IN` keyword in a SQL `WHERE` clause).

# Some context about the bento_lib query language
The query language is based on an AST written as nested lists such as: 
```
["#eq", ["#resolve", "patient", "name"], "John Doe"]
```
The first item in an array is always a function name which defines an `Expression`. By convention, they start with a pound sign. The list of valid function names is defined in  ./search/queries.py

An Expression can be composed of nested Expressions, or Literals. In the above example, the "root" expression starting by `#eq` contains a nested Expression starting by `#resolve` and a Literal `"John Doe"`
The number or arguments a given Expression can have is defined by its range property. The mapping from the functions names to their range values is encoded in .search/queries.py: `FUNCTION_ARGUMENTS`

The query can be transpiled to SQL using the psycopg2 library. The result is a psycopg2 SQL object and its corresponding parameters (you can think about this as a prepared statement where the SQL commands are separated from the interpolated values). It is not straigthforward to see the resulting SQL string as it requires a database cursor (i.e. an actual database connection to Postgresql, and a cursor from that database) to use the `cursor.mogrify()` command from psycopg2 (nonetheless, it is doable knowing the postgreSQL port, and the DB credentials).
The code for the pyscopg2 evaluation is in ./search/postgre.py

The functions in that file correspond to each query Expression function (for example `#contains` and `contains`). Each return the two parts required by psychopg2: the SQL object and the parameters of the query. Each is also a recursive function that calls the evaluation of its nested arguments into psycopg2 constructs.

The query can also be evaluated "in memory", that is using python function on actual results. An example of this is the federation service where a part of the query construct is executed in memory when performing joins between results sets. The code for this is in ./search/data_structure.py

# Design decisions
My initial take on this feature was to just add an `#in` operator that would take an Expression resolving to a field (`["#resolve", "table_name", "[item]", "field_name"]`) as a first argument and a list of literals as its second argument.
Due to the way the query string is parsed to generate the AST, it is expected that a list is a valid Expression, that is, it must start with a keyword. In other words, it can't be a plain list of literals.

Due to this constraint, I've added a `#list` function name to implement a syntax for expressing a list of literals.

This keeps the `#in` operator consistent with the other binary operators such as `#eq`, `#contains`, etc... which all takes only 2 arguments.

# How to test
- There is a unit test in the test suite that can be run with tox. Note that a part of the test suite is testing the events module of the library and relies on a connection to redis with a hardcoded local URL. I've dismissed that part of the test suite to test locally
- If you wish to see the resulting SQL, you'd need to setup a connection to a postgreSQL server. I did that suring development using the dev version of the katsu-db container which has an open port. The credentials are available in the env variables for bentoV2.
Another approach is to use the `features/leaner-queries` branch in Katsu project which uses the version 2.3 of Bento-lib, that contains the same changes. Running `python manage.py test chord_metadata_services/chord/tests/test_search.py` calls 3 unit test that use the `#in` and `#list` constructs.
